### PR TITLE
koka: init at 2.1.1

### PR DESCRIPTION
**Pull request by siraben**
###### Motivation for this change
Add Koka compiler https://github.com/koka-lang/koka

```ShellSession
$ koka /tmp/koka/samples/basic/caesar.kk 
compile: /tmp/koka/samples/basic/caesar.kk
loading: std/core
loading: std/core/types
loading: std/core/hnd
loading: std/num/double
loading: std/text/parse
loading: std/num/int32
check  : tmp/koka/samples/basic/caesar
linking: tmp_koka_samples_basic_caesar
created: out/v2.1.1/clang-debug/tmp_koka_samples_basic_caesar
 
plain  : Koka is a well-typed language
encoded: Krnd lv d zhoo-wbshg odqjxdjh
cracked: Koka is a well-typed language
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 117447` at c80387e9 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>koka</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

cmake is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/haskell-modules/generic-builder.nix:308:3:
```
    |
308 |   buildInputs = otherBuildInputs ++ optionals (!isLibrary) propagatedBuildInputs;
    |   ^
```
</li>
  </ul>
</details>

**Comment by siraben** [comment]
Never mind, that was in my overlay, in this PR everything seems to be fine:

```ShellSession
$ nix build .#koka && ./result/bin/koka /tmp/koka/samples/basic/caesar.kk 
compile: /tmp/koka/samples/basic/caesar.kk
loading: std/core
loading: std/core/types
loading: std/core/hnd
compile: /nix/store/pyv9j1cr8ii8xvgkvdfnzr92q16dz315-koka-2.1.1/share/koka/v2.1.1/lib/std/num/double.kk
compile: /nix/store/pyv9j1cr8ii8xvgkvdfnzr92q16dz315-koka-2.1.1/share/koka/v2.1.1/lib/std/text/parse.kk
check  : std/text/parse
compile: /nix/store/pyv9j1cr8ii8xvgkvdfnzr92q16dz315-koka-2.1.1/share/koka/v2.1.1/lib/std/num/int32.kk
check  : std/num/int32
check  : std/num/double
check  : tmp/koka/samples/basic/caesar
linking: tmp_koka_samples_basic_caesar
ld: warning: -sdk_version and -platform_version are not compatible, ignoring -sdk_version
created: out/v2.1.1/clang-debug/tmp_koka_samples_basic_caesar
 
plain  : Koka is a well-typed language
encoded: Krnd lv d zhoo-wbshg odqjxdjh
cracked: Koka is a well-typed language
```

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 117447` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>koka</li>
  </ul>
</details>

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [inline pkgs/development/compilers/koka/default.nix:32]: You can remove all these and the inputs as well as I don't think we will be able to run tests anytime soon as they need stack.

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [inline pkgs/development/compilers/koka/default.nix:37]: You should probably wrap the executable here as well with its run time dependencies. I think we need at least `cmake`, `gcc` and related binutils. 

good way to test is to try and use `koka` inside a `nix-shell --pure -p koka -I nixpkgs=/path/to/nixpkgs`

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [inline pkgs/development/compilers/koka/default.nix:48]: Feel free to add me as well :)

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [reply pkgs/development/compilers/koka/default.nix:37]: Also be mindful that `stdenv` contains a C compiler in the nix-shell by default.

**Comment by siraben** [comment]
Running in pure Nix shell, found out I needed cmake and gcc in path.

```ShellSession
$ nix-shell --pure -p koka -I nixpkgs=. 
$ koka /tmp/koka/samples/basic/caesar.kk 
compile: /tmp/koka/samples/basic/caesar.kk
loading: std/core
loading: std/core/types
loading: std/core/hnd
loading: std/num/double
loading: std/text/parse
loading: std/num/int32
check  : tmp/koka/samples/basic/caesar
linking: tmp_koka_samples_basic_caesar
created: out/v2.1.1/clang-debug/tmp_koka_samples_basic_caesar
 
plain  : Koka is a well-typed language
encoded: Krnd lv d zhoo-wbshg odqjxdjh
cracked: Koka is a well-typed language
```

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 117447` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>koka</li>
  </ul>
</details>

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [inline pkgs/development/compilers/koka/default.nix:21]: Can you check if `stdenv.cc` works here instead if you do `koka.override { stdenv = clangStdenv; }` would be pretty cool if it could support clang.

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [inline pkgs/development/compilers/koka/default.nix:40]: Turns out, not necessary for `wrapProgram`:

<https://github.com/NixOS/nixpkgs/blob/0c51d091cb187512f2408ae9c41afb496e1c9404/pkgs/build-support/setup-hooks/make-wrapper.sh#L145>

File: pkgs/development/compilers/koka/default.nix
Comment (siraben) [reply pkgs/development/compilers/koka/default.nix:21]: Yeah, works here since on macOS it uses clangStdenv.

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [inline pkgs/development/compilers/koka/default.nix:39]: Okay, so `stdenv.cc` is wrong, since it isn't suitable for cross, my bad. We want a C compiler for the buildPlatform and not take the C compiler of the hostPlatform.

Also for koka to properly pick up on the C compiler, we need to `--set CC` in `wrapProgram` which is actually kind of neat.

OTOH statically setting a C compiler may not the best idea since users may want to choose different ones, for example by setting `CC`?? Easiest would probably to statically take `gcc` as an input, I'm not sure.

Comment (siraben) [reply pkgs/development/compilers/koka/default.nix:39]: I could set it to buildPackages.stdenv.cc`

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [inline pkgs/development/compilers/koka/default.nix:39]: `koka` checks `CC` for the C compiler and prefers gcc over clang if it is in `PATH` somewhere else, so we need to set `CC`.

```suggestion
    wrapProgram "$out/bin/koka" \
      --set CC "${lib.getBin buildPackages.stdenv.cc}/bin/${buildPackages.stdenv.cc.targetPrefix}cc" \
```

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [inline pkgs/development/compilers/koka/default.nix:21]: having the C compiler in here is not necessary technically if you set `CC`.

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/koka/default.nix
Comment (sternenseemann) [reply pkgs/development/compilers/koka/default.nix:39]: Seems overrideable enough, would be especially neat if we had `pkgsClang` :|

**Review by sternenseemann — APPROVED** [review]
Kinda wondering if `pkgsCross.*.koka` actually works, but I would've to bootstrap GHC to find out…

**Comment by SuperSandro2000** [comment]
> Kinda wondering if `pkgsCross.*.koka` actually works, but I would've to bootstrap GHC to find out…

let me test that quickly.

**Comment by SuperSandro2000** [comment]
and after almost an hour it build.

**Comment by sternenseemann** [comment]
> and after almost an hour it build.

Can you try starting it and typing something like `1 + 1` and pressing enter? That should invoke the C compiler and produce a result (yes koka is at that stage where the REPL generates C, compiles and executes it…).

**Comment by SuperSandro2000** [comment]
> yes koka is at that stage where the REPL generates C, compiles and executes it…

I only tested it with binfmt and after waiting a few seconds it showed me 2. 

```
compile: /nix/store/f68vrp2q5hyrj6n7fph4aji5z7jswwzw-koka-2.1.1-aarch64-unknown-linux-gnu/share/koka/v2.1.1/lib/std/core.kk                                                   compile: /nix/store/f68vrp2q5hyrj6n7fph4aji5z7jswwzw-koka-2.1.1-aarch64-unknown-linux-gnu/share/koka/v2.1.1/lib/std/core/types.kk                                             check  : std/core/types
compile: /nix/store/f68vrp2q5hyrj6n7fph4aji5z7jswwzw-koka-2.1.1-aarch64-unknown-linux-gnu/share/koka/v2.1.1/lib/std/core/hnd.kk
check  : std/core/hnd
check  : std/core

> 1+1
check  : interactive
check  : interactive
cmake  : kklib from: /nix/store/f68vrp2q5hyrj6n7fph4aji5z7jswwzw-koka-2.1.1-aarch64-unknown-linux-gnu/share/koka/v2.1.1/kklib
-- The C compiler identification is GNU 10.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/hotpi/src/nixpkgs4/out/v2.1.1/cc-debug/kklib
Scanning dependencies of target kklib
[  7%] Building C object CMakeFiles/kklib.dir/src/bits.c.o
[ 14%] Building C object CMakeFiles/kklib.dir/src/box.c.o
[ 21%] Building C object CMakeFiles/kklib.dir/src/bytes.c.o
[ 28%] Building C object CMakeFiles/kklib.dir/src/init.c.o
[ 35%] Building C object CMakeFiles/kklib.dir/src/integer.c.o
[ 42%] Building C object CMakeFiles/kklib.dir/src/os.c.o
[ 50%] Building C object CMakeFiles/kklib.dir/src/process.c.o
[ 57%] Building C object CMakeFiles/kklib.dir/src/random.c.o
[ 64%] Building C object CMakeFiles/kklib.dir/src/refcount.c.o
[ 71%] Building C object CMakeFiles/kklib.dir/src/ref.c.o
[ 78%] Building C object CMakeFiles/kklib.dir/src/string.c.o
[ 85%] Building C object CMakeFiles/kklib.dir/src/time.c.o
[ 92%] Building C object CMakeFiles/kklib.dir/mimalloc/src/static.c.o
[100%] Linking C static library libkklib.a
[100%] Built target kklib
linking: interactive
created: out/v2.1.1/cc-debug/interactive
```

The resulting binaries in $PWD/out are x86-out though. This could be the cause of binfmt though.

**Comment by siraben** [comment]
@SuperSandro2000 Maybe we could address that in a future PR

**Comment by sternenseemann** [comment]
> The resulting binaries in $PWD/out are x86-out though

Might be that koka only targets `x86` and always defaults to `x86_64`, but would have dig to the code a bit more to confirm. <https://github.com/koka-lang/koka/blob/e330ca7c835f9f899134451e5d710420f7572362/src/Compiler/Options.hs#L259>

**Comment by siraben** [comment]
Modulo that, anything else blocking merge?

File: pkgs/top-level/all-packages.nix
Comment (anka-213) [inline pkgs/top-level/all-packages.nix:10593]: We should probably change this to:
```
  koka = haskell.lib.justStaticExecutables (haskellPackages.callPackage ../development/compilers/koka { });
```
so
1. We don't needlessly include all the haskell library dependencies in the closure
and
2. Users don't get a bunch of
```
error: packages '/nix/store/someProgram/lib/links/libgmpxx.4.dylib' and '/nix/store/someOtherProgram/lib/links/libgmpxx.4.dylib' have the same priority 5; use 'nix-env --set-flag priority NUMBER INSTALLED_PKGNAME' to change the priority of one of the conflicting packages (0 being the highest priority)
```
errors when trying to install it on MacOS.

Comment (anka-213) [reply pkgs/top-level/all-packages.nix:10593]: I should probably open a separate pull request for this though, since I was too late to catch this before it was merged.

Comment (anka-213) [reply pkgs/top-level/all-packages.nix:10593]: With vs without `justStaticExecutables`:
```
❯ nix path-info -Sh /nix/store/vs77d70z834c3hvzz73a87sl5xkb0w65-koka-2.1.1/
/nix/store/vs77d70z834c3hvzz73a87sl5xkb0w65-koka-2.1.1   817.0M
❯ nix path-info -Sh /nix/store/5mshznqzns5lndkv4632m886dwlvkbka-koka-2.1.1/
/nix/store/5mshznqzns5lndkv4632m886dwlvkbka-koka-2.1.1     2.4G
```

Comment (sternenseemann) [reply pkgs/top-level/all-packages.nix:10593]: Yeah, good point.

Comment (anka-213) [reply pkgs/top-level/all-packages.nix:10593]: I've created a pr now: https://github.com/NixOS/nixpkgs/pull/119857

**Pull request merged**
